project/Component/NefCompiler/Instances/NefCompilerSystem.swift

```diff
@@ -81,19 +87,39 @@ class NefCompilerSystem: CompilerSystem {
             return EnvIO.pure(frameworks)^
         }
         
+        func extractSwiftModules(paths: [String]) -> EnvIO<CompilerSystemEnvironment, CompilerSystemError, [SwiftModule]> {
+            EnvIO.access { fileSystem -> [String] in
+                let directories = paths.filter(fileSystem.isDirectory)
+                return directories.filter { $0.filename.extension == "swiftmodule" }
+            }
+             .flatMap { modules in modules.swiftModules() }^
+             .contramap(\.fileSystem)
+        }
+        
         func copyFrameworks(paths: [String], nefPlayground: NefPlaygroundURL) -> EnvIO<CompilerSystemEnvironment, CompilerSystemError, Void> {
             EnvIO { env in
                 env.fileSystem.copy(itemPaths: paths, to: nefPlayground.appending(.fw).path).void()
             }.mapError { _ in .build(nefPlayground.project, info: "move frameworks into '\(nefPlayground.project.path)'") }
         }
         
+        func copySwiftModules(_ swiftModules: [SwiftModule], nefPlayground: NefPlaygroundURL) -> EnvIO<CompilerSystemEnvironment, CompilerSystemError, Void> {
+            EnvIO { env in
+                let copyModules = env.fileSystem.copy(itemPaths: swiftModules.map(\.module).map(\.path), to: nefPlayground.appending(.swiftmodules).path)
```

Comment:
I think you can avoid one `map` here by using `.map(\.module.path)`, avoiding going over the array twice.

Reply:
ahahah, you are right!